test/celeritas/optical/ValidationUtils.hh

```diff
@@ -0,0 +1,136 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/ValidationUtils.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <array>
+#include <vector>
+
+#include "corecel/cont/Array.hh"
+#include "corecel/cont/Span.hh"
+#include "corecel/data/Collection.hh"
+#include "celeritas/UnitTypes.hh"
+#include "celeritas/grid/GenericGridData.hh"
+#include "celeritas/io/ImportPhysicsVector.hh"
+#include "celeritas/optical/MfpBuilder.hh"
+
+#include "Test.hh"
+#include "TestMacros.hh"
+
+#include "detail/ValidationUtilsImpl.hh"
+
+//! Equivalence macro for physics grids
+#define EXPECT_GRID_EQ(expected, actual) \
+    EXPECT_PRED_FORMAT2(::celeritas::testdetail::IsGridEq, expected, actual)
+
+//! Equivalence macro for physics tables (vectors of grids)
+#define EXPECT_TABLE_EQ(expected, actual) \
+    EXPECT_PRED_FORMAT2(::celeritas::testdetail::IsTableEq, expected, actual)
+
+namespace celeritas
+{
+namespace optical
+{
+namespace test
+{
+using namespace ::celeritas::test;
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to annotate units of a hard-coded test data array.
+ *
+ * Converts the arguments supplied in units \c UnitType to native units.
+ */
+template<class UnitType, class... Args>
+Array<real_type, sizeof...(Args)> constexpr native_array_from(Args const&... args)
+{
+    return Array<real_type, sizeof...(Args)>{
+        native_value_from(UnitType(args))...};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to annotate units of a hard-coded test data array.
+ *
+ * This overload accepts fixed size \c Arrays which clang-format handles
+ * better.
```

Comment:
This (and the detail) are also neat, but I think it should be removed even if the format is ugly (which we can always suppress with the `// clang-format: off`) for two reasons:
1. The manually specified `N` will silently make the array bigger and fill it with zeros if it's greater than the number of actual entries given
2. You have to go through several extra hoops (forward declaring the function in the impl file) to get it to work.

So I'd say just kill this function and replace its single use in RayleighMfp with the one above.

Reply:
P.S. I know I was the one who recommended the trailing comma to you (but at the time I didn't see that it was a function call rather than brace initialization). So I'm sorry about the misdirection!

Reply:
All good - when I wrote this it had a bit of code smell to me, and just manually formatting is definitely more appropriate.